ARROW-3701: [Gandiva] add op for decimal 128

pravindra opened the pull request:
The code changes are complete. However, the perf in the non-fast code path is slow - I'll debug and fix that.

wesm commented on cpp/src/gandiva/CMakeLists.txt:
```diff
@@ -146,6 +150,7 @@ if (ARROW_GANDIVA_BUILD_TESTS)
   add_gandiva_unit_test(lru_cache_test.cc)
   add_gandiva_unit_test(to_date_holder_test.cc to_date_holder.cc date_utils.cc)
   add_gandiva_unit_test(simple_arena_test.cc)
+  add_gandiva_unit_test(decimal_type_util_test.cc decimal_type_util.cc)
```
I'm curious why not just link the test files with libgandiva.a? This is similar to the `ADD_PARQUET_TEST` macro we have here https://github.com/apache/arrow/blob/master/cpp/src/parquet/CMakeLists.txt#L28

wesm replied:
Sure, no problem

pitrou commented on cpp/src/gandiva/decimal_type_util.h:
```diff
@@ -0,0 +1,123 @@
+  };
+
+  /// The maximum precision representable by a 4-byte decimal
+  static constexpr int32_t kMaxDecimal4Precision = 9;
```
I would call those "Decimal64" and "Decimal128" out of consistency.

pravindra replied:
will fix.

pitrou commented on cpp/src/gandiva/decimal_type_util.h:
```diff
@@ -0,0 +1,123 @@
+  /// The maximum precision representable by a 16-byte decimal
+  static constexpr int32_t kMaxPrecision = 38;
+
+  static constexpr int32_t kMaxScale = kMaxPrecision;
```
I don't understand this. Why is the max scale equal to the max precision?

pravindra replied:
sorry, I didn't get the question - the max possible scale and max possible precision are both 38.

pitrou replied:
I see. Is that checked on inputs? In Arrow, we don't seem to have a limitation on Decimal scale or precision.

pravindra replied:
not yet. I'll add a TODO, and add all the validations in a follow-up PR.

pitrou commented on cpp/src/gandiva/decimal_type_util.h:
```diff
@@ -0,0 +1,123 @@
+  static constexpr int32_t kMaxPrecision = 38;
+
+  static constexpr int32_t kMaxScale = kMaxPrecision;
+  static constexpr int32_t kMinAdjustedScale = 6;
```
What is this?

pravindra replied:
will add a comment. The scale can be reduced upto a value of 6 if the increase precision exceeds 38.

pravindra replied:
no specific reason - both sql-server and impala use 6 as the minimum. So, I picked 6 too.

pitrou commented on cpp/src/gandiva/decimal_type_util.h:
```diff
@@ -0,0 +1,123 @@
+};
+
+inline Decimal128TypePtr DecimalTypeUtil::MakeType(int32_t precision, int32_t scale) {
+  return std::make_shared<arrow::Decimal128Type>(precision, scale);
```
You can use the `arrow::decimal()` factory function (see `type.h`).

pravindra replied:
that is returning a base DataType and so, I was having to cast it.

pitrou commented on cpp/src/gandiva/decimal_type_util_test.cc:
```diff
@@ -0,0 +1,61 @@
+      DoOp(DecimalTypeUtil::kOpAdd, MakeDecimal(38, 10), MakeDecimal(38, 38)));
+
+  EXPECT_ARROW_TYPE_EQUALS(
+      MakeDecimal(38, 6),
```
Why? Doesn't this create truncation if the 2-scale decimal is too large? I would expect `decimal(38, 1)`.

pravindra replied:
the minimum scale is 6 post-adjustment (kMinAdjustedScale).

pitrou commented on cpp/src/gandiva/function_ir_builder.cc:
```diff
@@ -0,0 +1,59 @@
+
+llvm::Value* FunctionIRBuilder::BuildIfElse(llvm::Value* condition,
+                                            llvm::Type* return_type,
+                                            std::function<llvm::Value*()> then_func,
```
I'm curious: what does it bring to pass a function here, rather than directly the corresponding `llvm::Value`?

pravindra replied:
you are right - I'll fix this.

pravindra replied:
sorry, I agreed too soon - these have to be functions. The code-generation for the if/else blocks must happen after the corresponding BasicBlock has started i.e after the SetInsertPoint call. Otherwise, there is no conditional evaluation i.e both if/else blocks will get evaluated.

Tried the following simple case:
   _x_value = BuildIfElse(lt_max_precision, Add11(x_value), Sub11(x_value));_

**without the functions (i.e using direct lvalue, I get)**
```
if:
  %3 = add i128 %member, 11
  %4 = sub i128 %member, 11
  br i1 %2, label %then, label %else

exit:                                             ; preds = %merge
  ret void

then:                                             ; preds = %entry
  br label %merge

else:                                             ; preds = %entry
  br label %merge

merge:                                            ; preds = %else, %then
  %res_value = phi i128 [ %3, %then ], [ %4, %else ]
```
**with the lambda functions**

 ```
 %2 = icmp slt i32 %member1, 38
  br i1 %2, label %then, label %else

exit:                                             ; preds = %merge
  ret void

then:                                             ; preds = %entry
  %3 = add i128 %member, 11
  br label %merge

else:                                             ; preds = %entry
  %4 = sub i128 %member, 11
  br label %merge

merge:                                            ; preds = %else, %then
  %res_value = phi i128 [ %3, %then ], [ %4, %else ]
```

pitrou commented on cpp/src/gandiva/function_ir_builder.cc:
```diff
@@ -0,0 +1,59 @@
+  auto then_value = then_func();
+  builder->CreateBr(merge_bb);
+
+  // refresh then_bb for phi (could have changed due to code generation of then_vv).
```
Surely you mean "then_value".

pravindra replied:
no, I actually mean then_bb. basically, I want to refresh the "current block" which could have changed due to code-generation (maybe, the then_func has an if-else block too).

pitrou commented on cpp/src/gandiva/function_ir_builder.cc:
```diff
@@ -0,0 +1,59 @@
+  auto else_value = else_func();
+  builder->CreateBr(merge_bb);
+
+  // refresh else_bb for phi (could have changed due to code generation of else_vv).
```
"else_value"

pitrou commented on cpp/src/gandiva/function_ir_builder.cc:
```diff
@@ -0,0 +1,59 @@
+
+namespace gandiva {
+
+llvm::Value* FunctionIRBuilder::BuildIfElse(llvm::Value* condition,
```
Isn't this function just reimplementing the `select` instruction? https://llvm.org/docs/LangRef.html#select-instruction

pravindra replied:
no - select-instruction is only conditional assignment. It doesn't do control flow.

For the same example as above, with select (instead of if-else), I get 

  ```
%2 = icmp slt i32 %member1, 38
  %3 = add i128 %member, 11
  %4 = sub i128 %member, 11
  %5 = select i1 %2, i128 %3, i128 %4
```

pitrou commented on cpp/src/gandiva/arrow.h:
```diff
@@ -48,6 +51,14 @@ using ArrayDataVector = std::vector<ArrayDataPtr>;
 
+static inline bool is_decimal_128(DataTypePtr type) {
+  if (type->id() == arrow::Type::DECIMAL) {
+    auto decimal_type = dynamic_cast<arrow::DecimalType*>(type.get());
```
You can use `arrow::internal::checked_cast` (see `util/checked_cast.h`).

pravindra replied:
will fix.

pitrou commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,470 @@
+    return AddFastPath(x_value, x_scale, y_value, y_scale);
+  };
+  auto else_lambda = [&] {
+    // slower version : add whole/fraction parts separately
```
I don't understand this: when is the slow path needed? When you have full precision?

pravindra replied:
yes - when, the result precision is = 38.

pitrou commented on cpp/src/gandiva/llvm_generator.cc:
```diff
@@ -389,6 +402,43 @@ llvm::Value* LLVMGenerator::AddFunctionCall(const std::string& full_name,
+  llvm::IRBuilder<>* builder = ir_builder();
+
+  // only decimals of size 128-bit supported.
+  DCHECK_EQ(is_decimal_128(arrow_type), true);
```
`DCHECK(is_decimal_128(arrow_type))`

pitrou commented on cpp/src/gandiva/llvm_generator.cc:
```diff
@@ -953,25 +1021,45 @@ LValuePtr LLVMGenerator::Visitor::BuildValueAndValidity(const ValueValidityPair&
+  if (arrow_return_type_id == arrow::Type::DECIMAL) {
+    // For decimals, the output struct pointer is passed along as an input
+    // param to the function call. This allows for a stack allocation instead
+    // of a malloc/new.
```
I'm not sure what you mean. AFAIK you should be able to return a struct type from a LLVM function.

pravindra replied:
but, that would be an extra copy. And, I still need to pass along the result precision/scale to the LLVM function (I don't want it to recompute for each record).

pravindra replied:
Ok. The choice is between 

(a) struct_decimal result = add(struct_decimal x, struct_decimal y, int32_t result_precision, int32_t result_scale)

vs

(b) void add(const struct_decimal *x, const struct_decimal *y, struct_decimal *result)

I like b better :-).

pravindra replied:
do let me know if you prefer (a) strongly - I'll switch to (a).

pitrou replied:
No, it's fine.

pitrou commented on cpp/src/gandiva/llvm_types.h:
```diff
@@ -125,6 +147,20 @@ class LLVMTypes {
+  };
+
+  llvm::StructType* decimal128_struct_type() {
+    return llvm::StructType::get(context_, {i128_type(), i32_type(), i32_type()}, false);
```
This looks a bit wasteful. The types are always going to be the same along a single batch.

pitrou replied:
This is a struct for a decimal128 _value_, right? Meaning you pass a separate scale and precision for each value?

pravindra replied:
okay, I get what you are saying. I can separate out the scale/precision to a different structure, that is inited only once. 

I didn't realize this since the optimization passes inlined the entire add function, and removed the structs during my testing. but, I think it's better to not depend on that.

I'll fix this.

pitrou commented:
I'm skeptical about the semantics here. If adding two decimals you're increasing the scale and precision. But that's as if adding two `int32` gave a `int33`. The usual idiom with decimals seems to keep a fixed precision accross all computations (see e.g. Python's [decimal](https://docs.python.org/3/library/decimal.html) library).

pravindra commented:
> I'm skeptical about the semantics here. If adding two decimals you're increasing the scale and precision. But that's as if adding two `int32` gave a `int33`. The usual idiom with decimals seems to keep a fixed precision accross all computations (see e.g. Python's [decimal](https://docs.python.org/3/library/decimal.html) library).

This is not true for SQL implementations, including Impala.

https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-2017

https://docs.oracle.com/javadb/10.6.2.1/ref/rrefsqlj36146.html#rrefsqlj36146

Generally, for addition, the scale remains the same, but precision increases by 1 (to accommodate the extra digit).

pravindra commented:
@pitrou - can you please review this change again ?

There are some significant changes since the last iteration.
1.  addressed the review comments
2. re-orged the IR code so that the fast-paths stay in decimal-ir.cc, but the more complex ones go into cpp code (decimal.cc or decimal_sql.cc), which can be compiled into ir-helpers.bc
3. the cpp impl (decimal_sql.cc) now optimises for the case where the numbers aren't big enough to cause overflow on addition (i.e there are leading zeroes).

With the tests added in gandiva benchmarks, the results are (for 1M additions on a single thread)

Benchmark                                        Time           CPU Iterations
DecimalAdd2Fast/min_time:1.000                1896 us       1895 us        735
DecimalAdd2LeadingZeroes/min_time:1.000       4170 us       4162 us        324
DecimalAdd2Large/min_time:1.000              44610 us      44544 us         30
DecimalAdd3Fast/min_time:1.000                2661 us       2660 us        502
DecimalAdd3LeadingZeroes/min_time:1.000       9384 us       9370 us        139
DecimalAdd3Large/min_time:1.000             106828 us     106661 us         13

wesm commented:
I believe @pitrou is away until the new year, if you don't mind waiting until then

pravindra commented:
@wesm @praveenbingo can you please review instead ?

wesm commented:
I can review but it won't be until the day after Christmas. It would be good for @pitrou to leave comments anyway if this gets merged later this week

pravindra commented:
No problem @wesm. Thanks.

pitrou commented on cpp/src/gandiva/lvalue.h:
```diff
@@ -30,17 +32,48 @@ class LValue {
   llvm::Value* validity_;
 };
 
+class LValueDecimal : public LValue {
```
Call this `DecimalLValue`?

pravindra replied:
fixed

pitrou commented on cpp/src/gandiva/llvm_generator.cc:
```diff
@@ -1007,12 +1080,9 @@ std::vector<llvm::Value*> LLVMGenerator::Visitor::BuildParams(
-      params.push_back(result_ref.length());
-    }
+    // append all the parameters corresponding to this LValue.
+    result_ref.AppendFunctionParams(&params);
```
Though it seems the front-end (i.e. us) is supposed to be responsible for the complicated parts of the ABI when talking to LLVM :-/
e.g. https://stackoverflow.com/questions/25127874/c-abi-with-llvm and 
https://stackoverflow.com/questions/22776391/why-does-clang-coerce-struct-parameters-to-ints

pravindra replied:
yup, sometime back, I tried this and hit the coerce issue with structs. So, reverted to passing the primitive params.

pitrou commented on cpp/src/gandiva/precompiled/decimal_ops.cc:
```diff
@@ -0,0 +1,41 @@
+  gandiva::Decimal128Full y(y_high, y_low, y_precision, y_scale);
+
+  arrow::Decimal128 out = gandiva::decimalsql::Add(x, y, out_precision, out_scale);
+  return DecimalSplit{out.high_bits(), out.low_bits()};
```
I wonder how ABI-proof it is to return a struct here... (see comments above about LLVM not handling some parts of the ABI)

pravindra replied:
from what i understood, it works fine for trivial structs (no constructors). seems to work fine in our tests.

is there an alternative ?

pravindra replied:
ah.. ok. It does look ugly, I've just added a comment for now.

pitrou commented on cpp/src/gandiva/precompiled/decimal_sql.h:
```diff
@@ -0,0 +1,34 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Can you explain why this is called `decimal_sql.h`? Does it have anything to do with SQL?

pravindra replied:
i was initially thinking we may want to support decimal ops where the sql rules but scale/precision don't matter (eg. like python add). but, probably we can get to it when required.

for now, i have removed the 'sql' from the names.

pitrou commented on cpp/src/gandiva/precompiled/decimal_sql.h:
```diff
@@ -0,0 +1,34 @@
+namespace gandiva {
+namespace decimalsql {
+
+Decimal128 Add(const Decimal128Full& x, const Decimal128Full& y, int32_t out_precision,
```
Add a docstring? Is this function solely meant to test the algorithm implemented in LLVM IR elsewhere?

pravindra replied:
added docstring. used in testing and the entry-point for the IR function (from add_large_decimal128_decimal128 in decimal_wrapper.cc)

pitrou commented on cpp/src/gandiva/precompiled/decimal_sql_test.cc:
```diff
@@ -0,0 +1,71 @@
+  Decimal128TypePtr out_type;
+  auto status =
+      DecimalTypeSql::GetResultType(DecimalTypeSql::kOpAdd, {t1, t2}, &out_type);
+  EXPECT_TRUE(status.ok()) << status.message();
```
ASSERT_OK

pravindra replied:
fixed

pitrou commented on cpp/src/gandiva/precompiled/decimal_sql_test.cc:
```diff
@@ -0,0 +1,71 @@
+               Decimal128Full{"100", 38, 7},                                     // y
+               Decimal128Full{"99999999999999999999999999999990000010", 38, 6});
+
+  // Both -ve
```
Also add a test for negative and max prec?

pravindra replied:
done

pitrou commented on cpp/src/gandiva/tests/CMakeLists.txt:
```diff
@@ -27,11 +27,17 @@ ADD_GANDIVA_TEST(to_string_test)
 ADD_GANDIVA_TEST(in_expr_test)
 ADD_GANDIVA_TEST(null_validity_test)
+ADD_GANDIVA_TEST(decimal_test)
+ADD_GANDIVA_TEST(decimal_ops)
```
All other tests end with `_test`. Why deviate?

pravindra replied:
fixed

pitrou commented on cpp/src/gandiva/tests/CMakeLists.txt:
```diff
@@ -27,11 +27,17 @@ ADD_GANDIVA_TEST(to_string_test)
   USE_STATIC_LINKING)
 
-ADD_BENCHMARK(micro_benchmarks
+ADD_GANDIVA_TEST(decimal_ops_static
```
Same here.

pravindra replied:
fixed

pitrou commented on cpp/src/gandiva/tests/decimal_ops.cc:
```diff
@@ -0,0 +1,214 @@
+  // Evaluate expression
+  arrow::ArrayVector outputs;
+  status = projector->Evaluate(*in_batch, pool_, &outputs);
+  EXPECT_TRUE(status.ok()) << status.message();
```
`ASSERT_OK`

pravindra replied:
fixed

pitrou commented on cpp/src/gandiva/tests/decimal_ops.cc:
```diff
@@ -0,0 +1,214 @@
+               decimal_literal("2001099", 38, 6));
+
+  AddAndVerify(decimal_literal("1000999", 38, 6),   // x
+               decimal_literal("10000992", 38, 7),  // y
```
What happens with `10000995` here?

pravindra replied:
added test

pitrou commented on cpp/src/gandiva/tests/decimal_ops.cc:
```diff
@@ -0,0 +1,214 @@
+               decimal_literal("-2001099", 38, 6));
+
+  AddAndVerify(decimal_literal("-1000999", 38, 6),   // x
+               decimal_literal("-10000992", 38, 7),  // y
```
Same question here with `-10000995`.

pravindra replied:
added test

pitrou commented on cpp/src/gandiva/tests/decimal_test.cc:
```diff
@@ -0,0 +1,237 @@
+    EXPECT_TRUE(status.ok()) << status.message();
+
+    Decimal128 scaled_value;
+    status = str_value.Rescale(str_scale, scale, &scaled_value);
```
Please use `DCHECK_OK` here.

pravindra replied:
done

pravindra commented on cpp/src/arrow/util/decimal-test.cc:
```diff
@@ -466,4 +466,73 @@ TEST(Decimal128Test, TestToInteger) {
+  ASSERT_OK(fraction.ToInteger(&out));
+  ASSERT_EQ(6, out);
+
+  value.GetWholeAndFraction(5, &whole, &fraction);
```
done

pravindra commented on cpp/src/arrow/util/decimal-test.cc:
```diff
@@ -466,4 +466,73 @@ TEST(Decimal128Test, TestToInteger) {
+  ASSERT_OK(result.ToInteger(&out));
+  ASSERT_EQ(-1238, out);
+
+  result = Decimal128("-123749").ReduceScaleBy(2, true);
```
becomes -1238 (>= 10^2/2). added test.

pitrou replied:
Ok. AFAIK, round-to-even is generally preferred for accuracy, but we can change this later.

pravindra commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -139,9 +139,24 @@ class ARROW_EXPORT Decimal128 {
+  /// \brief Scale up.
+  Decimal128 IncreaseScaleBy(int32_t increase_by) const;
+
+  /// \brief Scale down and round.
```
if 'round' is false, the extra digits are simply discarded. no rounding is done. updated comment.

pravindra commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -139,9 +139,24 @@ class ARROW_EXPORT Decimal128 {
+  /// \brief Scale down and round.
+  Decimal128 ReduceScaleBy(int32_t reduce_by, bool round = true) const;
+
+  /// \brief count the number of leading zeroes.
```
changed fn name and comment.

pravindra commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -884,6 +925,54 @@ Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
 
+void Decimal128::GetWholeAndFraction(int scale, Decimal128* whole,
+                                     Decimal128* fraction) const {
+  Decimal128 multiplier(ScaleMultipliers[scale]);
```
done

pravindra commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -884,6 +925,54 @@ Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
+}
+
+const Decimal128& Decimal128::GetScaleMultiplier(int32_t scale) {
+  return ScaleMultipliers[scale];
```
done

pravindra commented on cpp/src/gandiva/decimal_full.h:
```diff
@@ -0,0 +1,74 @@
+
+using Decimal128 = arrow::Decimal128;
+
+/// Represents a 128-bit decimal value along with it's precision and scale.
```
fixed

pravindra commented on cpp/src/gandiva/decimal_full.h:
```diff
@@ -0,0 +1,74 @@
+  Decimal128Full(int32_t precision, int32_t scale)
+      : value_(0), precision_(precision), scale_(scale) {}
+
+  inline bool Equals(const Decimal128Full& o) const {
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+    ValueSplit(llvm::Value* high, llvm::Value* low) : high_(high), low_(low) {}
+
+    // Make from i128 value
+    static ValueSplit Make(DecimalIR* decimal_ir, llvm::Value* in);
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+    // Make from i128 value
+    static ValueSplit Make(DecimalIR* decimal_ir, llvm::Value* in);
+
+    // Make from IR struct
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+    static ValueSplit MakeFromIR(DecimalIR* decimal_ir, llvm::Value* dstruct);
+
+    // Combine the two parts into an i128
+    llvm::Value* Combine(DecimalIR* decimal_ir) const;
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+    static ValueWithOverflow MakeFromIR(DecimalIR* decimal_ir, llvm::Value* dstruct);
+
+    // Build a corresponding IR struct
+    llvm::Value* BuildIRStruct(DecimalIR* decimal_ir) const;
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+        : value_(value), overflow_(overflow) {}
+
+    // Make from IR struct
+    static ValueWithOverflow MakeFromIR(DecimalIR* decimal_ir, llvm::Value* dstruct);
```
done

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+  // Add global variables to the module.
+  static void AddGlobals(Engine* engine);
+
+  // Get intrinsic functions that are used by decimal operations.
```
fixed

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+                                         llvm::Value* out_scale);
+
+  // Do addition of large integers (both positive and negative).
+  llvm::Value* AddLarge(llvm::Value* x_value, llvm::Value* x_precision,
```
Added ValueFull - has value, scale and precision.

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+  void AddTrace128(const std::string& msg, llvm::Value* value);
+
+  // name of the global variable having the array of scale multipliers.
+  static const char kScaleMultipliersName[];
```
fixed

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+// CPP:  x <= y ? y : x
+llvm::Value* DecimalIR::GetHigherScale(llvm::Value* x_scale, llvm::Value* y_scale) {
+  llvm::Value* le = ir_builder()->CreateICmpSLE(x_scale, y_scale);
+  return BuildIfElse(le, types()->i32_type(), [&] { return y_scale; },
```
switched to select.

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+
+  // trace strings - these need to be kept around since the execution happens much
+  // after the code build.
+  std::vector<std::unique_ptr<uint8_t>> trace_strings_;
```
removed trace_strings_.

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+  //                           int128_t y_value, int32_t y_precision, int32_t y_scale
+  //                           int32_t out_precision, int32_t out_scale)
+  std::vector<llvm::Type*> arguments;
+  arguments.push_back(types()->i128_type());
```
added a BuildFunction() that takes tuples of type/name.

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+                  Status::CodeGenError("Error creating function."));
+  // Name the arguments
+  llvm::Function::arg_iterator args = function->arg_begin();
+  llvm::Value* x_value = &*args;
```
added a BuildFunction() that takes tuples of type/name.

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+      BuildIfElse(lt_max_precision, types()->i128_type(), then_lambda, else_lambda);
+
+  // store result to out
+  ir_builder()->CreateBr(exit);
```
fixed

pravindra commented on cpp/src/gandiva/decimal_type_sql.cc:
```diff
@@ -0,0 +1,29 @@
+
+namespace gandiva {
+
+constexpr int32_t DecimalTypeSql::kMaxDecimal32Precision;
```
tried removing them - linker complains about undefined symbols.

pitrou replied:
Ah, ok. My bad.

pravindra commented on cpp/src/gandiva/decimal_type_sql.h:
```diff
@@ -0,0 +1,129 @@
+  static Decimal128TypePtr MakeAdjustedType(int32_t precision, int32_t scale);
+};
+
+inline Decimal128TypePtr DecimalTypeSql::MakeType(int32_t precision, int32_t scale) {
```
I need the return type to be Decimal128Type. I modified the fn to use arrow::decimal, and type-cast.

pravindra commented on cpp/src/gandiva/decimal_type_sql.h:
```diff
@@ -0,0 +1,129 @@
+}
+
+// Implementation of decimal rules.
+inline Status DecimalTypeSql::GetResultType(Op op, const Decimal128TypeVector& in_types,
```
moved to .cc

pravindra commented on cpp/src/gandiva/decimal_type_sql.h:
```diff
@@ -0,0 +1,129 @@
+// Implementation of decimal rules.
+inline Status DecimalTypeSql::GetResultType(Op op, const Decimal128TypeVector& in_types,
+                                            Decimal128TypePtr* out_type) {
+  // TODO : validations
```
done

pravindra commented on cpp/src/gandiva/decimal_type_sql_test.cc:
```diff
@@ -0,0 +1,61 @@
+                       Decimal128TypePtr d2) {
+  Decimal128TypePtr ret_type;
+  auto status = DecimalTypeSql::GetResultType(op, {d1, d2}, &ret_type);
+  EXPECT_TRUE(status.ok()) << status.message();
```
fixed

pravindra commented on cpp/src/gandiva/decimal_type_sql_test.cc:
```diff
@@ -0,0 +1,61 @@
+namespace gandiva {
+
+Decimal128TypePtr MakeDecimal(int32_t precision, int32_t scale) {
+  return std::make_shared<arrow::Decimal128Type>(precision, scale);
```
fixed

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+void DecimalIR::AddTrace(const std::string& fmt, std::vector<llvm::Value*> args) {
+  DCHECK(enable_ir_traces_);
+
+  // The string needs to be saved for later execution.
```
thanks for pointing this out - fixed.

pravindra commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -0,0 +1,461 @@
+    AddTrace128(msg, value);      \
+  }
+
+const char DecimalIR::kScaleMultipliersName[] = "kScaleMultipliers";
```
added gandiva prefix

pravindra commented on cpp/src/gandiva/engine.h:
```diff
@@ -68,6 +70,10 @@ class Engine {
   void AddGlobalMappingForFunc(const std::string& name, llvm::Type* ret_type,
                                const std::vector<llvm::Type*>& args, void* func);
 
+  void AddFunctionIRBuilder(std::shared_ptr<FunctionIRBuilder> func_ir_builder) {
```
removed this.

pravindra commented on cpp/src/gandiva/function_ir_builder.h:
```diff
@@ -0,0 +1,63 @@
+  Engine* engine_;
+};
+
+// TODO
```
removed them - not required any longer.

pravindra commented on cpp/src/gandiva/function_registry.cc:
```diff
@@ -38,6 +38,7 @@ using arrow::utf8;
 using std::vector;
 
 #define STRINGIFY(a) #a
+#define decimal128() arrow::decimal(0, 0)
```
this file is an exception - we use the typename to generate the  signatures in the macros with STRINGIFY (eg. add_decimal128_decimal128). Added a comment.

pravindra commented on cpp/src/gandiva/function_signature.h:
```diff
@@ -56,10 +56,26 @@ class FunctionSignature {
+          // For decimal types, the precision/scale isn't part of the signature.
+          auto dleft = arrow::internal::checked_cast<arrow::DecimalType*>(left.get());
+          auto dright = arrow::internal::checked_cast<arrow::DecimalType*>(right.get());
+          match = (dleft != NULL) && (dright != NULL) &&
```
done

pravindra commented on cpp/src/gandiva/function_signature.h:
```diff
@@ -56,10 +56,26 @@ class FunctionSignature {
+          break;
+        }
+        default:
+          match = left->Equals(right);
```
done

pravindra commented on cpp/src/gandiva/literal_holder.h:
```diff
@@ -22,11 +22,14 @@
 namespace gandiva {
 
 using LiteralHolder =
     boost::variant<bool, float, double, int8_t, int16_t, int32_t, int64_t, uint8_t,
```
is it ok to use a different jira/PR for this ? This PR has already become too big.

pitrou replied:
Yes, it's ok :-)

pitrou commented:
Thanks for addressing the comments. I'll take a last quick look tomorrow.

pitrou commented on cpp/src/gandiva/to_date_holder_test.cc:
```diff
@@ -62,6 +62,10 @@ TEST_F(TestToDateHolder, TestSimpleDateTime) {
       to_date(&execution_context, "1986-12-11 01:30:00", true, &out_valid);
   EXPECT_EQ(millis_since_epoch, 534643200000);
+
+  millis_since_epoch =
```
What this added to this PR by mistake?

pravindra replied:
yes, removed now.

pitrou commented on cpp/src/gandiva/tests/micro_benchmarks.cc:
```diff
@@ -289,5 +399,13 @@ BENCHMARK(TimedTestFilterLike)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
 BENCHMARK(TimedTestAllocs)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
 BENCHMARK(TimedTestMultiOr)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
 BENCHMARK(TimedTestInExpr)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
+BENCHMARK(DecimalAdd2Fast)->MinTime(1.0)->Unit(benchmark::kMicrosecond);
```
At some point you may want to split the micro-benchmarks into several files.

pitrou commented on cpp/src/gandiva/precompiled/decimal_wrapper.cc:
```diff
@@ -0,0 +1,43 @@
+  uint64_t low_bits;
+};
+
+FORCE_INLINE
```
By curiosity, is this picked up by LLVM when compiling IR?
Generally, I would trust LLVM to inline small functions when a high optimization level is selected.

pravindra replied:
early on when I was adding gandiva functions, I remember this made a difference. but, it's likely I didn't have all the optimizer passes working at that point. I'll test again and check if this is still required.

pravindra commented:
@wesm @pitrou can you please merge this change ?

The CI tests pass, and I've addressed all the review comments.

wesm commented on cpp/src/gandiva/tests/decimal_test.cc:
```diff
@@ -0,0 +1,237 @@
+      MakeArrowArrayDecimal(decimal_type, MakeDecimalVector({"2", "3", "4", "5"}, scale),
+                            {true, true, true, true});
+
+  auto array_c = MakeArrowArrayBool({true, false, true, false}, {true, true, true, true});
```
This should probably be replaced with `ArrayFromJSON`

pravindra replied:
makes sense. but this is a big change, deserves a new jira !

wesm commented on cpp/src/gandiva/tests/decimal_test.cc:
```diff
@@ -0,0 +1,237 @@
+  auto node_c = TreeExprBuilder::MakeField(field_c);
+  auto if_node = TreeExprBuilder::MakeIf(node_c, node_a, node_b, decimal_type);
+
+  auto expr = TreeExprBuilder::MakeExpression(if_node, field_result);
```
Thought: might be nice to have some more concise aliases, like `gandiva::expr(gandiva::if_(gandiva::field(...))` or something

pravindra replied:
i was hoping ARROW-3458 would make writing tests much easier. but, I agree your suggestion is a much easier way forward.

wesm commented on cpp/src/gandiva/tests/generate_data.h:
```diff
@@ -79,6 +81,24 @@ class Int64DataGenerator : public DataGenerator<int64_t> {
+ protected:
+  bool large_;
+  Random random_;
+};
```
Let us try to prioritize consolidating our random data generation code

pravindra replied:
will do

wesm commented on cpp/src/gandiva/tests/test_util.h:
```diff
@@ -46,6 +47,14 @@ static ArrayPtr MakeArrowArray(std::vector<C_TYPE> values) {
+  ArrayPtr out;
+  arrow::ArrayFromVector<TYPE, C_TYPE>(type, validity, values, &out);
+  return out;
+}
```
Put this in `arrow/test-util.h`?

pravindra replied:
will do.

wesm commented on cpp/src/gandiva/tests/test_util.h:
```diff
@@ -68,11 +77,16 @@ static ArrayPtr MakeArrowTypeArray(const std::shared_ptr<arrow::DataType>& type,
 
+#define EXPECT_ARROW_TYPE_EQUALS(a, b)                                \
+  EXPECT_TRUE((a)->Equals(b)) << "expected type: " << (a)->ToString() \
+                              << " actual type: " << (b)->ToString();
```
A lot of these defines (or some version of them) should be consolidated

pravindra replied:
will do

wesm commented:
I'll work on getting my Windows branch ready to go so we can begin to maintain the build on Windows

pravindra commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -884,6 +925,54 @@ Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
+  DCHECK_GE(reduce_by, 0);
+  DCHECK_LE(reduce_by, 38);
+
+  Decimal128 divisor(ScaleMultipliers[reduce_by]);
```
is it ok to add as needed ?

pravindra commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -139,9 +139,24 @@ class ARROW_EXPORT Decimal128 {
+  /// \brief Scale down and round.
+  Decimal128 ReduceScaleBy(int32_t reduce_by, bool round = true) const;
+
+  /// \brief count the number of leading zeroes.
```
I think you are looking at an old version. In the latest change, this is called CountLeadingBinaryZeros

pravindra commented on cpp/src/gandiva/decimal_full.h:
```diff
@@ -0,0 +1,74 @@
+using Decimal128 = arrow::Decimal128;
+
+/// Represents a 128-bit decimal value along with it's precision and scale.
+class Decimal128Full {
```
agree, will change.

pravindra commented on cpp/src/gandiva/decimal_full.h:
```diff
@@ -0,0 +1,74 @@
+  Decimal128Full(int64_t high_bits, uint64_t low_bits, int32_t precision, int32_t scale)
+      : value_(high_bits, low_bits), precision_(precision), scale_(scale) {}
+
+  Decimal128Full(std::string value, int32_t precision, int32_t scale)
```
will fix.

pravindra commented on cpp/src/gandiva/decimal_full.h:
```diff
@@ -0,0 +1,74 @@
+      : value_(value), precision_(precision), scale_(scale) {}
+
+  Decimal128Full(int32_t precision, int32_t scale)
+      : value_(0), precision_(precision), scale_(scale) {}
```
the default constructor for Decimal128 sets the value to 0. so, i'll remove the value_(0).

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+// under the License.
+
+#ifndef GANDIVA_DECIMAL_ADD_IR_BUILDER_H
+#define GANDIVA_DECIMAL_ADD_IR_BUILDER_H
```
will fix

pravindra commented on cpp/src/gandiva/decimal_ir.h:
```diff
@@ -0,0 +1,164 @@
+  // Holder for an i128 value, and a boolean indicating overflow.
+  class ValueWithOverflow {
+   public:
+    ValueWithOverflow(llvm::Value* value, llvm::Value* overflow)
```
will fix

pravindra commented on cpp/src/gandiva/function_registry.cc:
```diff
@@ -38,6 +38,7 @@ using arrow::utf8;
 using std::vector;
 
 #define STRINGIFY(a) #a
+#define decimal128() arrow::decimal(0, 0)
```
i think you are looking at a stale change. The latest one has this moved as an inline function to function_registry_common.h (commented as internal only).

we use this for generating the fn names in the macros (eg. add_decimal128_decimal128).

pravindra commented on cpp/src/gandiva/function_signature.h:
```diff
@@ -56,10 +56,26 @@ class FunctionSignature {
+          break;
+        }
+        default:
+          match = left->Equals(right);
```
this is already fixed in the latest change :-)

pravindra commented:
@wesm, thanks for the review. I'll address the easy ones in my next PR for decimal. And, open new jiras for the rest.
